src/main.cpp

```diff
@@ -5806,6 +5808,218 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
+                percentage,
+                subtreeIndex,
+                chainSubtreeIndex
+            );
+            percentage += 10;
+        }
+
+        int nHeight = vHeights[subtreeIndex];
+
+        auto pindex = chainActive[nHeight];
+        CBlock block;
+        if (!ReadBlockFromDisk(block, pindex, consensusParams)) {
+            LogPrintf("Failed to read block\n");
+            return false;
+        }
+
+        // We'll grab the final frontier from the previous block (which
+        // should have a hashFinalSaplingRoot/hashFinalOrchardRoot
+        // because this block completed a 2^16 size subtree!) and append
+        // to it until we complete the subtree.
+        if (type == SAPLING) {
+            SaplingMerkleTree sapling_tree;
+            assert(pcoinsTip->GetSaplingAnchorAt(pindex->pprev->hashFinalSaplingRoot, sapling_tree));
+            for (const CTransaction &tx : block.vtx) {
+                for (const auto &outputDescription : tx.GetSaplingOutputs()) {
+                    sapling_tree.append(uint256::FromRawBytes(outputDescription.cmu()));
+
+                    auto completeSubtreeRoot = sapling_tree.complete_subtree_root();
+                    if (completeSubtreeRoot.has_value()) {
+                        libzcash::SubtreeData subtree(completeSubtreeRoot->ToRawBytes(), nHeight);
+                        pcoinsTip->PushSubtree(SAPLING, subtree);
+                        goto nextsubtree; // sorry
+                    }
+                }
+            }
+
+            // We should not get here; this block should have completed the subtree
+            // and the goto statement above should have executed.
+            assert(false);
+        } else if (type == ORCHARD) {
+            OrchardMerkleFrontier orchard_tree;
+            assert(pcoinsTip->GetOrchardAnchorAt(pindex->pprev->hashFinalOrchardRoot, orchard_tree));
+            for (const CTransaction &tx : block.vtx) {
+                if (tx.GetOrchardBundle().IsPresent()) {
+                    try {
+                        auto appendResult = orchard_tree.AppendBundle(tx.GetOrchardBundle());
+                        if (appendResult.has_subtree_boundary) {
+                            libzcash::SubtreeData subtree(appendResult.completed_subtree_root, nHeight);
+
+                            pcoinsTip->PushSubtree(ORCHARD, subtree);
+                            goto nextsubtree; // sorry
+                        }
+                    } catch (const rust::Error& e) {
+                        return false;
+                    }
+                }
+            }
+
+            // Similarly, we should not get here.
+            assert(false);
+        } else {
+            assert(false);
+        }
+
+        nextsubtree:
+            // needed because statements need to exist
+            // after a label is defined or it won't parse
+            continue;
```

Comment:
```suggestion
        auto pushSaplingSubtrees = [&]() {
            SaplingMerkleTree sapling_tree;
            assert(pcoinsTip->GetSaplingAnchorAt(pindex->pprev->hashFinalSaplingRoot, sapling_tree));
            for (const CTransaction &tx : block.vtx) {
                for (const auto &outputDescription : tx.GetSaplingOutputs()) {
                    sapling_tree.append(uint256::FromRawBytes(outputDescription.cmu()));

                    auto completeSubtreeRoot = sapling_tree.complete_subtree_root();
                    if (completeSubtreeRoot.has_value()) {
                        libzcash::SubtreeData subtree(completeSubtreeRoot->ToRawBytes(), nHeight);
                        pcoinsTip->PushSubtree(SAPLING, subtree);
                        return;
                    }
                }
            }
        };
        auto pushOrchardSubtrees = [&]() {
            OrchardMerkleFrontier orchard_tree;
            assert(pcoinsTip->GetOrchardAnchorAt(pindex->pprev->hashFinalOrchardRoot, orchard_tree));
            for (const CTransaction &tx : block.vtx) {
                if (tx.GetOrchardBundle().IsPresent()) {
                    auto appendResult = orchard_tree.AppendBundle(tx.GetOrchardBundle());
                    if (appendResult.has_subtree_boundary) {
                        libzcash::SubtreeData subtree(appendResult.completed_subtree_root, nHeight);
                        pcoinsTip->PushSubtree(ORCHARD, subtree);
                        return;
                    }
                }
            }
        };

        if (type == SAPLING) {
            pushSaplingSubtrees();
        } else if (type == ORCHARD) {
            try {
                pushOrchardSubtrees();
            } catch (const rust::Error& e) {
                return false;
            }
        } else {
            assert(false);
        }
```
(The scope of the `try` in the Orchard case is wider with this approach. Nevertheless I think it's cleaner.)